src/modules/claim/services/claim-issuance.service.ts

```diff
@@ -0,0 +1,150 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import jwt from 'jsonwebtoken';
+import { IClaimIssuance, DecodedClaimToken } from '../claim.types';
+import { RoleService } from '../../role/role.service';
+import { Logger } from '../../logger/logger.service';
+import { ClaimIssueDTO, NewClaimIssueDTO } from '../claim.dto';
+import { RoleClaim } from '../entities/roleClaim.entity';
+import { ClaimService } from './claim.service';
+import { ClaimHandleResult } from '../claim-handle-result.dto';
+
+@Injectable()
+export class ClaimIssuanceService {
+  constructor(
+    private readonly roleService: RoleService,
+    private readonly logger: Logger,
+    @InjectRepository(RoleClaim)
+    private readonly roleClaimRepository: Repository<RoleClaim>
+  ) {
+    this.logger.setContext(ClaimIssuanceService.name);
+  }
+
+  /**
+   * Handles claim issuance request saving and updates.
+   * Two scenarios are handled - issue requested claim and issue not-requested claim
+   * @param rq IClaimIssuance request
+   */
+  public async handleClaimIssuanceRequest(
+    rq: IClaimIssuance,
+    previouslyRequestedClaim: RoleClaim | undefined
+  ): Promise<ClaimHandleResult> {
+    if (!previouslyRequestedClaim && (rq.issuedToken || rq.onChainProof)) {
+      return this.handleClaimWithoutPreviousRequest(rq);
+    } else if (
+      previouslyRequestedClaim &&
+      !previouslyRequestedClaim.isRejected &&
+      (rq.issuedToken || rq.onChainProof)
+    ) {
+      return this.handleClaimWithPreviousRequest(rq, previouslyRequestedClaim);
+    }
+
+    return ClaimHandleResult.Failure('claim issuance criteria not met');
+  }
+
+  private async handleClaimWithoutPreviousRequest(
+    rq: IClaimIssuance
+  ): Promise<ClaimHandleResult> {
+    let claimType: string;
+    let claimTypeVersion: string;
+
+    if (rq.issuedToken) {
```

Comment:
Probably this can be omitted as claim issued without request should include `claimType` and `claimTypeVersion` https://github.com/energywebfoundation/iam-client-lib/blob/2ff0fa4c4a2d44cf13f8a919352932791154ebac/src/modules/claims/claims.service.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L361

Reply:
Currently is just for On-chain claims
https://github.com/energywebfoundation/iam-client-lib/blob/2ff0fa4c4a2d44cf13f8a919352932791154ebac/src/modules/claims/claims.service.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L391

Reply:
Ah, right. I guess will need to add it to off-chain claim as well

Reply:
It will clarify logic a little bit for sure